compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -796,6 +797,140 @@ static FailureOr<Flow::WorkloadBuilder> getWorkloadBuilder(OpBuilder &builder,
+    });
+  };
+  auto isSingleLoop = [&](iterator slowIt, iterator fastIt) {
+    return slowIt + 1 == fastIt;
+  };
+
+  Optional<iterator> slowIt = llvm::None;
+  if (isAllAffineContiguous(parallelDims.begin()))
+    slowIt = parallelDims.begin();
+  for (iterator fastIt = parallelDims.begin() + 1; fastIt != parallelDims.end();
+       fastIt++) {
+    if (slowIt.has_value() &&
+        (!isAllAffineContiguous(fastIt) || !isLoopContiguous(fastIt))) {
+      if (!isSingleLoop(slowIt.value(), fastIt))
+        parallelContiguousLoops.push_back({slowIt.value(), fastIt});
+      slowIt = llvm::None;
+    }
+    if (slowIt.has_value() && !isAllAffineContiguous(fastIt))
+      slowIt = llvm::None;
+    if (!slowIt.has_value() && isAllAffineContiguous(fastIt)) {
+      slowIt = fastIt;
+    }
+  }
+  if (slowIt.has_value() && (slowIt.value() + 1) != parallelDims.end())
+    parallelContiguousLoops.push_back({slowIt.value(), parallelDims.end()});
+
+  return parallelContiguousLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic and return the new
+/// one
+static FailureOr<linalg::GenericOp> collapseLinalgGeneric(
+    TensorDimTrackingRewriter &rewriter, linalg::GenericOp genericOp) {
+  SmallVector<unsigned int> parallelDims;
+  genericOp.getParallelDims(parallelDims);
+  SmallVector<ReassociationIndices> collapseIndices =
+      getParallelContiguousLoops(genericOp, parallelDims);
+
+  rewriter.setInsertionPoint(genericOp);
+  Optional<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  // Find and return collapsed linalg.generic
+  if (auto expandshapeOp =
+          replacements->front().getDefiningOp<tensor::ExpandShapeOp>()) {
+    if (auto newGenericOp =
+            expandshapeOp.getOperand().getDefiningOp<linalg::GenericOp>()) {
+      rewriter.replaceOp(genericOp, *replacements);
+      return newGenericOp;
+    }
+  }
+
+  return failure();
+}
+
+/// Return the given op is collapsable
+static bool isEligibleForCollapse(Operation *op,
+                                  ArrayRef<Operation *> producers) {
+  // No producers atm
+  if (!producers.empty()) return false;
+
+  // Collapse only linalg.generics atm
+  if (!isa<linalg::GenericOp>(op)) return false;
+
+  auto genericOp = dyn_cast<linalg::GenericOp>(op);
+  // TODO Cannot generate tensor.collapse_shape with dynamic shapes
+  for (Value output : genericOp.getOutputs()) {
+    if (auto tensorType = dyn_cast<TensorType>(output.getType())) {
+      if (tensorType.getNumDynamicDims() > 0) {
+        return false;
+      }
+    }
+  }
+
+  // Need at least 3D+
+  if (kNumMaxParallelDims >=
+      (pruneUnitTripLoops(genericOp.getStaticLoopRanges())))
```

Comment:
I dont think we need to make a distinction between unit-trip loops?

Reply:
Right, I removed that.